druid/src/core.rs

```diff
@@ -579,6 +585,43 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
         }
     }
 
+    fn update_enabled(
```

Comment:
Thanks for the review :)

Hmm, i haven't thought about that at all.
I see the point. It could get messy if we just send the event during another one.

I can think of two options to solve this:
- First: we add a fourth bool, something like `child-enabled-changed` and add a `InternalLifeCycle::RouteEnabledChanged()` event which gets send down the tree if `child-enabled-changed` is set or enabled was changed for the current widget. Then we could either update the focus chain directly or use children_changed.
- Second: we generalize `InternalLifeCycle::RouteWidgetAdded` to `InternalLifeCycle::RouteWidgetChanged` which would perform both tasks at once. This should work, since a widget can't receive both (`WidgetAdded` and `EnabledChanged`) at the same time. I like this version better since we save another `InternalLifeCycle` event, won't pollute the widgetstate with further dirty flags and also it feels like both events behave quite similar, since they are the only events which change the focus-chain. But i think for this to work, widgets should not be allowed to call `enabled_changed` from `Lifecycle::WidgetAdded`.

What are your thoughts about this?

Reply:
The second option hadn't occurred to me, but I think it's *possible* it could work? I'd want to really think it through. One interesting thing about this approach is that we could potentially re-use the `children_changed` flag to indicate where we need to recurse in order to deliver the `FocusChanged` event.

Thinking this through more, though, I have my doubts. Basically: I think that this is exchanging superficial complexity ("one less event type! one less flag in `WidgetState`) for potentially deep combinational complexity, e.g. "now these two distinct actions interact in ways that are hard to reason about". For instance: imagine if, in response to the same event, one widget changes a child, and another widget calls `set_disabled`. Now we want to send `WidgetAdded`, but in the course of doing this we reset `children_changed`, and how we can't correctly route the `FocusChanged` event? I'm not sure if this is an actual concern; the point is more that I'm already getting a bit confused thinking about it.

So I think I prefer the first option. It's more verbose, but it is also simpler to reason about. Adding an additional `Internal` event really doesn't cost us anything meaningful.

A separate question is, in this world, what state we need to store in `WidgetState`. My current understanding is that we will need, at least (and without committing to these names, I'm just using what comes to me right now):

- `is_self_disabled`: whether *this specific widget* has been disabled
- `is_disabled`: whether this widget is part of a disabled subtree (including the case where it is a subtree of one; `is_self_disabled` implies `is_disabled`).
- `child_disabled_changed` a flag used when routing the actual event; when this is `false` we can stop recursing.
- `disabled_changed`: a flag which indicates that the *current* widget has been disabled. Actually: I think maybe this isn't a bool, but is more like an `Option<bool>`, to indicate the new state?

Oh no, this gets worse.

What happens when a widget and one of its descendants both change their disabled state in response to the same event, but they do it in different directions? For instance: what if a subtree was disabled, and the parent re-enables it, and at the same time one of the children explicitly disables itself?

Maybe this isn't that bad. With the actual `EnabledChanged` event, we recurse down, stopping at any child that has either `is_self_disabled` set, or that has `disabled_changed` as `Some(true)`, or whatever thing means "I'm becoming disabled"? In this latter case we would need to continue to descend only if the change was to also become disabled; otherwise we just set `is_self_disabled`, for that particular widget, but continue to notify children that they *are* now disabled.

So..... I think this is *okay* (in that it is not fundamentally broken) but it is definitely more complicated than I had initially thought, and I suspect there are edge cases I'm not thinking of yet. I would encourage you to try and think through these various permutations of possible states and orders, and I *definitely* think we are going to want some tests in place, if just to check that our understanding is correct.

Reply:
I think the way you described should work, but i will definitely write some testcases this time :)

Reply:
But i'm unsure about one thing:

> `is_disabled`: whether this widget is part of a disabled subtree (including the case where it is a subtree of one; `is_self_disabled` implies `is_disabled`)

My current Implementation has two fields: one for the own state (manually disabled) and one for the tree (ancestors, not including my self, disabled). I think we need both, since otherwise we lose the information if calling `set_enabled()` will change the "enabled state" of this widget.

Also `new_disabled` feels more robust than `enabled_changed`. The question is, is it possible to receive an event between calling `set_enabled()` and receiving `LifeCycle::EnabledChanged()` and if so should this event use the old or the new state?

Reply:
> But i'm unsure about one thing:
> 
> > `is_disabled`: whether this widget is part of a disabled subtree (including the case where it is a subtree of one; `is_self_disabled` implies `is_disabled`)
> 
> My current Implementation has two fields: one for the own state (manually disabled) and one for the tree (ancestors, not including my self, disabled). I think we need both, since otherwise we lose the information if calling `set_enabled()` will change the "enabled state" of this widget.

I agree that we need two flags, one for the specific widget and one that includes parent state, and overrides the state of the specific widget; if a parent is disabled but we're not, we still *are*. I think we're on the same page about this.

> Also `new_disabled` feels more robust than `enabled_changed`. 

If you can make this work with `new_disabled` that's great. I guess maybe how I was thinking about this is as a two-part process; in the first part (when the widget calls `set_disabled()` we store information on *the pending change*, and then in the second part (when we traverse the tree to reconcile the changes and send the events) we clear the *pending* state and set the *actual* state. So in my thinking a widget isn't disabled when it calls `set_disabled`; it is disabled when the tree is traversed afterwards.

> The question is, is it possible to receive an event between calling `set_enabled()` and receiving `LifeCycle::EnabledChanged()` and if so should this event use the old or the new state?

The only event that it should be possible to receive here is `RouteWidgetAdded`; we need to always handle that before we do anything else. I think it should be fine to keep the new state in place while handling this event; immediately afterwards we will handle enabled/disabled, and then after that (if necessary) we will handle focus change.